src/System.Text.Json/src/System/Text/Json/JsonUtf8Reader.cs

```diff
@@ -0,0 +1,1274 @@
+                }
+            }
+
+            Debug.Assert(nextByte == 'E' || nextByte == 'e');
+            i++;
+
+            signResult = ConsumeSign(ref data, ref i, nextByte);
+            if (signResult == ConsumeNumberResult.NeedMoreData)
+                return false;
+
+            Debug.Assert(signResult == ConsumeNumberResult.OperationIncomplete);
+
+            i++;
+            ConsumeNumberResult resultExponent = ConsumeIntegerDigits(ref data, ref i);
+            if (resultExponent == ConsumeNumberResult.NeedMoreData)
+                return false;
+            if (resultExponent == ConsumeNumberResult.Success)
+                goto Done;
+            if (resultExponent == ConsumeNumberResult.OperationIncomplete)
+            {
+                _lineBytePosition += i;
+                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedEndOfDigitNotFound, nextByte);
+            }
+
+        Done:
+            ValueSpan = data.Slice(0, i);
+            consumed = i;
+            return true;
+        }
+
+        private ConsumeNumberResult ConsumeNegativeSign(ref ReadOnlySpan<byte> data, ref int i)
+        {
+            byte nextByte = data[i];
+
+            if (nextByte == '-')
+            {
+                i++;
+                if (i >= data.Length)
+                {
+                    if (IsLastSpan)
+                    {
+                        _lineBytePosition += i;
+                        ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedDigitNotFoundEndOfData, nextByte);
+                    }
+                    else
+                        return ConsumeNumberResult.NeedMoreData;
+                }
+
+                nextByte = data[i];
+                if ((uint)(nextByte - '0') > '9' - '0')
+                {
+                    _lineBytePosition += i;
+                    ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedDigitNotFound, nextByte);
+                }
+            }
+            return ConsumeNumberResult.OperationIncomplete;
+        }
+
+        private ConsumeNumberResult ConsumeZero(ref ReadOnlySpan<byte> data, ref int i)
+        {
+            i++;
+            byte nextByte = default;
+            if (i < data.Length)
+            {
+                nextByte = data[i];
+                if (JsonConstants.Delimiters.IndexOf(nextByte) >= 0)
+                    return ConsumeNumberResult.Success;
+            }
+            else
+            {
+                if (IsLastSpan)
+                    return ConsumeNumberResult.Success;
+                else
+                    return ConsumeNumberResult.NeedMoreData;
+            }
+            nextByte = data[i];
+            if (nextByte != '.' && nextByte != 'E' && nextByte != 'e')
+            {
+                _lineBytePosition += i;
+                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedNextDigitComponentNotFound, nextByte);
```

Comment:
I think we discussed this in person and I never wrote it down: For anywhere that keeps ThrowHelper.Throw as a thrower, please add a `Debug.Fail("ThrowHelper should have thrown");` after it.  (Though my preference, of course, is to throw local instead of in a helper except for where needs must.)

Reply:
I understand the motivation, but I don't think we should clutter the code like that.  We don't do it elsewhere, e.g. in corelib, where we use ThrowHelper extensively, and I don't think we should do it here.  I'd rather the motivation be addressed by pushing for the C# compiler feature that allows for methods to be annotated in some fashion as non-returning.

Reply:
> Every ThrowHelper can uniqely be "oops, I didn't throw".

Then that's a bug in ThrowHelper and should be fixed.

> I guess it's just my burden to have to stop every time and think about it.

Yes.  That's often the case with perf-vs-usability trade-offs; you have to think more about the correctness of the code written for performance because it's often more complex / more code / more nuances / etc.  I see this as just one case of that.  Thankfully, we have a convention: things named ThrowHelper always throw, so just add it to your mental list along with "return", "throw", etc. :)

Reply:
Conclusion: Not adding Debug.Fail to reduce the potential clutter.